granary/meetup.py

```diff
@@ -0,0 +1,102 @@
+# coding=utf-8
+"""Meetup.com source class.
+"""
+
+from . import source
+import logging
+from oauth_dropins.webutil import util
+import urllib.parse, urllib.request
+
+API_BASE = 'https://api.meetup.com'
+API_RSVPS = '/%(urlname)s/events/%(event_id)s/rsvps'
+
+class Meetup(source.Source):
+
+    DOMAIN = 'meetup.com'
+    NAME = 'Meetup.com'
+
+    def __init__(self, access_token):
+        self.access_token = access_token
+        pass
+
+    def create(self, obj, preview=False, include_link=source.OMIT_LINK, ignore_formatting=False):
+        return self._create(obj, preview, include_link, ignore_formatting)
+
+    def post_rsvp(self, urlname, event_id, response):
+        url = API_BASE + API_RSVPS % {
+                'urlname': urlname,
+                'event_id': event_id,
+                }
+        params = 'response=%(response)s' % {
+                'response': response,
+                }
+        logging.debug('Creating RSVP=%(rsvp)s for %(urlname)s %(event_id)s' % {
+            'rsvp': response,
+            'urlname': urlname,
+            'event_id': event_id,
+            })
+        return util.urlopen(urllib.request.Request(
```

Comment:
reuse oauth_dropins.meetup.urlopen_bearer_token()?

Reply:
Looks like that code may only work for a GET, not a POST, though? :thinking: https://github.com/snarfed/oauth-dropins/blob/master/oauth_dropins/meetup.py#L42

Reply:
it will pass through data=..., which i think automatically makes urlopen() POST, right? if not, feel free to parameterize the method.

Reply:
When updating to use it, it now gives me some errors around setting up the mocks, would you be able to help? Python isn't my strong point!

```
======================================================================
FAIL: test_create_rsvp_yes_with_www (granary.tests.test_meetup.MeetupTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/local3/lib/python3.7/site-packages/mox3/mox.py", line 2139, in new_method
    func(self, *args, **kwargs)
  File "/app/granary/tests/test_meetup.py", line 66, in test_create_rsvp_yes_with_www
    created = self.meetup.create(rsvp)
  File "/app/granary/meetup.py", line 33, in create
    return self._create(obj, False, include_link, ignore_formatting)
  File "/app/granary/meetup.py", line 92, in _create
    resp = self.post_rsvp(urlname, event_id, response)
  File "/app/granary/meetup.py", line 51, in post_rsvp
    return meetup.urlopen_bearer_token(url, self.access_token, data=params)
  File "/app/oauth_dropins/meetup.py", line 38, in urlopen_bearer_token
    return util.urlopen(urllib.request.Request(url, headers=headers), **kwargs)
  File "/app/oauth_dropins/webutil/util.py", line 1415, in urlopen
    return urllib.request.urlopen(url_or_req, *args, **kwargs)
  File "/app/local3/lib/python3.7/site-packages/mox3/mox.py", line 814, in __call__
    return mock_method(*params, **named_params)
  File "/app/local3/lib/python3.7/site-packages/mox3/mox.py", line 1113, in __call__
    expected_method = self._VerifyMethodCall()
  File "/app/local3/lib/python3.7/site-packages/mox3/mox.py", line 1180, in _VerifyMethodCall
    raise UnexpectedMethodCallError(self, expected)
mox3.mox.UnexpectedMethodCallError: Unexpected method call.  unexpected:-  expected:+
- function.__call__(<urllib.request.Request object at 0x7f3b306b4a90>, data=b'response=yes', timeout=15) -> None
+ function.__call__(<function TestCase.expect_urlopen.<locals>.check_request at 0x7f3b306b5b00>, timeout=15) -> <oauth_dropins.webutil.testutil.UrlopenResult object at 0x7f3b306b4ad0>

```

Reply:
oh looks like you maybe need to pass data=... to the Request constructor, not to urlopen()?

if that's not it, you may need to check for bytes value instead of str in the check fn, or pass str instead of bytes.

Reply:
OK I've raised https://github.com/snarfed/oauth-dropins/pull/26 as a dependency to this - once that's in we'll hopefully be able to use that call to `urlopen` and that'll work :+1: